File: solutions/zig/bottle-song/1/bottle_song.zig
const std = @import("std");

fn numberToWord(count: u32) []const u8 {
    return switch (count) {
        1 => "One",
        2 => "Two",
        3 => "Three",
        4 => "Four",
        5 => "Five",
        6 => "Six",
        7 => "Seven",
        8 => "Eight",
        9 => "Nine",
        10 => "Ten",
        else => unreachable,
    };
}

fn numberToWordLower(count: u32) []const u8 {
    return switch (count) {
        1 => "one",
        2 => "two",
        3 => "three",
        4 => "four",
        5 => "five",
        6 => "six",
        7 => "seven",
        8 => "eight",
        9 => "nine",
        10 => "ten",
        else => unreachable,
    };
}

pub fn recite(buffer: []u8, start_bottles: u32, take_down: u32) ![]const u8 {
    if (start_bottles == 0 or take_down == 0 or take_down > start_bottles) {
        return error.InvalidInput;
    }

    var writer = std.Io.Writer.fixed(buffer);

    const end_bottles = start_bottles - take_down;
    var current_bottle = start_bottles;

    while (current_bottle > end_bottles) {
        const bottles_asc = numberToWord(current_bottle);

        if (current_bottle < start_bottles) {
            try writer.writeByte('\n');
            try writer.writeByte('\n');
        }

        const bottle_str = if (current_bottle == 1) "bottle" else "bottles";

        try writer.print("{s} green {s} hanging on the wall,\n", .{ bottles_asc, bottle_str });
        try writer.print("{s} green {s} hanging on the wall,\n", .{ bottles_asc, bottle_str });

        try writer.print("And if one green bottle should accidentally fall,\n", .{});

        current_bottle -= 1;

        if (current_bottle == 0) {
            try writer.print("There'll be no green bottles hanging on the wall.", .{});
        } else {
            const next_word_lower = numberToWordLower(current_bottle);
            const next_bottle_str = if (current_bottle == 1) "bottle" else "bottles";

            try writer.print("There'll be {s} green {s} hanging on the wall.", .{ next_word_lower, next_bottle_str });
        }
    }
    return writer.buffered();
}

File: solutions/go/bottle-song/1/bottle_song.go
package bottlesong

import (
	"fmt"
)

func bottleNumber(bottle int) string {
	switch bottle {
	case 0:
		return "no"
	case 1:
		return "one"
	case 2:
		return "two"
	case 3:
		return "three"
	case 4:
		return "four"
	case 5:
		return "five"
	case 6:
		return "six"
	case 7:
		return "seven"
	case 8:
		return "eight"
	case 9:
		return "nine"
	case 10:
		return "ten"
	}
	return fmt.Sprintf("%d", bottle)
}

func bottleNumberCaps(bottle int) string {
	switch bottle {
	case 0:
		return "No"
	case 1:
		return "One"
	case 2:
		return "Two"
	case 3:
		return "Three"
	case 4:
		return "Four"
	case 5:
		return "Five"
	case 6:
		return "Six"
	case 7:
		return "Seven"
	case 8:
		return "Eight"
	case 9:
		return "Nine"
	case 10:
		return "Ten"
	}
	return fmt.Sprintf("%d", bottle)
}

func bottleName(bottle int) string {
	if bottle == 1 {
		return "bottle"
	}
	return "bottles"
}

func Recite(startBottles, takeDown int) []string {
	var result []string

	for i := 0; i < takeDown; i++ {
		current := startBottles - i
		next := current - 1

		t := fmt.Sprintf("%s green %s hanging on the wall,",
			bottleNumberCaps(current),
			bottleName(current))
		result = append(result, t)
		result = append(result, t)
		result = append(result, "And if one green bottle should accidentally fall,")
		t = fmt.Sprintf("There'll be %s green %s hanging on the wall.",
			bottleNumber(next),
			bottleName(next))
		result = append(result, t)
		if i < takeDown-1 {
			result = append(result, "")
		}
	}

	return result
}
